### DISCUSSION TITLE
Postgres ltree support / Open Source Contributions

### DISCUSSION THREAD

**Discussion by halfmatthalfcat** [prisma/prisma#12306] [Q&A]
Hey all, not sure if this is the right forum for this but wanted to start a conversation around adding [Postgres ltree](https://www.postgresql.org/docs/current/ltree.html) support into Prisma.

I've started the process of updating the chain of libraries in order for Prisma to fully support this new datatype.

1. [rust-postgres PR](https://github.com/sfackler/rust-postgres/pull/875)
2. [quaint working PR](https://github.com/halfmatthalfcat/quaint/pull/1)
3. prisma-engines PR is currently a work in progress (query, migration, etc)
4. Any further work in prisma/prisma for this to be officially supported (documentation, etc)

I'd like to gauge the Prisma team's (especially @janpio) at the feasibility of this being accepted into Prisma in the short term. I understand Prisma has it's own priorities but I'm sinking a lot of time into this and want to know whether it would be seriously considered for merge when it is ready (hopefully by the end of March). 

Appreciate any advice or guidance on how you've managed larger open source contributions to Prisma and what I should expect going forward. 

edit: There is an open issue for this https://github.com/prisma/prisma/issues/2568, however wanted to have more of a discussion on how this looks for me personally as an open source contributor.

**Comment by janpio** [comment]
Thanks for your work @halfmatthalfcat ❤️ 

Saw your email and was planning to answer next week, but here is fine as well. Actually better, as others can also chime in (and my "no work on the weekend rule is much laxer about spending time on GitHub 😆). I will open 2 "threads" here now: One to continue working on the feature, and another one to discuss the topic of Open Source contributions to Prisma.

This one is for discussing the feature:

The big open question we usually start with internally (after we decided it is worth spending our resources on a specific feature vs. on others), is how a database feature would look like in **Prisma Schema Language (PSL)** and how **support across databases** looks like. Anything that only works in one database is harder to integrate, unless it presents itself as a "native type" (where we can do `@db.Foo`, and just have to find a fitting parent type). So to have a conversation on how probable it is that we add something to Prisma, we should start there.

I know very little about `ltree`, so will now read a bit to learn the basics, and then come back to post my thoughts as a comment on this thread. Maybe we can come up with a description together how this would look like in PSL.

**Reply by halfmatthalfcat** [reply]
So I preliminarily reached out to @pimeys to discuss what it would look like in PSL. I was proposing a native `@db.Ltree` type for Postgres with a Prisma type as `String`. Internally, Postgres validates ltree, lquery and ltxtqueries conform to their respective patterns but they're essentially treated as plain strings and can be (de)serialized as such.

I agree that, since this is specific to Postgres, it does make the logistics in the codebase a bit more complex but as I've started integrating into `prisma-engines`, it's not as bad as I would have thought.

**Reply by janpio** [reply]
The code is usually not the biggest problem, the product design around it is the much bigger portion 🗡️ 
But native type and `String` as scalar type looks great - that is the simplest case for sure 👍 

My first learning from a bit of reading: `ltree` requires a **Postgres extension** https://www.postgresql.org/docs/current/ltree.html. That is always a bit problematic as  Prisma Migrations does not have great support for that yet (https://github.com/prisma/prisma/labels/topic%3A%20postgresql-extension) You can have them enabled by default on your database server (and on newly created databases for things like the shadow database), then it will usually be fine. You can add them manually to your migration files, and things should be fine as well - but the experience of people with that is pretty broad. 
Fortunately we are aware of that and are currently working on GIS support - which for PostgreSQL required the PostGIS extension. So I expect we will have a good story for managing extensions, especially for PostgreSQL, soon. (No issue weridly right now, will create on in a bit.) `@db.ltree` then would be the first native type that would depend on such an extension, so it might even be a good example to design this with.

**Reply by janpio** [reply]
We have the issue for PostgreSQL extension management now: https://github.com/prisma/prisma/issues/12308

I'll continue reading about ltree now - these additional functions might be interesting.

**Reply by janpio** [reply]
Next thing I learned is that you can use Gist indexes with `ltree` columns. There is an issue https://github.com/prisma/prisma/issues/10634, and we track them as part of our work on PostGIS obviously: https://github.com/prisma/prisma/issues/1798 This will need to be taken into consideration.

**Reply by janpio** [reply]
Continuing my reading:
- Main type is `ltree` which stores paths divided by `.`
- You can insert and manage path like data with Strings pretty easily
- There are new operators `@>` and `<@` which can be used to read data or delete data
- There are new functions `NLEVEL()` (count depth) and `SUBPATH()` (return... sub paths)
- There is a new operator `||` that can be used to concatenate paths

Also:
- There are also `lquery` and `ltxtquery`, but I am unclear how exactly those are to be used
- There are further operators `~`, `?`, `@` and different variations of these, but I am unclear how exactly those are to be used
- There are also further functions

That raises the questions: 
1. Would you want to work with the data as strings in Prisma Client as well? Or is there a better representation possible?
2. Where and how would you want to use the new `<@` and `@>` operators?
3. Where and how would you want to use the new functions?
4. Where and how would you use the new `||` concatenation operator?
5. The integration with GiST indexes seems pretty essential - what DX would you expect in Prisma for that?
6. How are other ORMs dealing with all this?

**Reply by halfmatthalfcat** [reply]
Glad you're digging in! Hope I can shed some light:

1. I believe working with strings is the easiest path to working with ltree, lquery and ltxtquery. It should be up to the user to create the well defined query rather than requiring an array and joining it within the client.
2. `<@` and `@>` are filter operators that seek whether a given `ltree` is an ancestor or descendant of another ltree. For instance:
    ```sql
    SELECT * FROM my_table WHERE path <@ 'a.b'
    ```
    would be looking for rows in `my_table` where the path column, a column of type `ltree`, contains values that are a descendant of `'a.b'`. Meaning if you have a row with a `path` of `'a.b.c'`, it would be returned. The reverse case for `@>`, where rows that are ancestors of `'a.b'` would be returned.

    Along these lines, `~` allows you to leverage `lquery` against a given ltree. `lquery` is basically a regex-like path specification where you can define more flexible querying semantics against `ltree`. For example:
    ```sql
    SELECT * FROM my_table WHERE path ~ 'a.*.c'
    ```
    Here we are looking for any row in `my_table` where the path column, an `ltree`, contains ltrees that begin with `a`, have anything as the second path part and then `c` in the third position. There are more advanced semantics to leverage with `lquery`, but that's the gist of it. Along those lines, `?` just allows you to filter against an array of lqueries for a given `ltree`. For example:
    ```sql
    SELECT * FROM my_table WHERE path ? ARRAY['a.*{0,2}.c', 'a.b|c.d@']::lquery[]
    ```
    This essentially says "Find all rows where path (1) starts with `a`, has a preceding 0 - 2 path parts (can be anything) and end in `c`, or (2) starts with `a`, has `b` or `c` as the second path part and has either `d` or `D` as the third.
    
    The `@` operator allows you to leverage `ltxtqueries` which are basically full-text searches of `ltree`s. For example:
    ```sql
    SELECT * FROM my_table WHERE path @ 'a | b & !c'
    ```
    Which reads as "Find all rows where an ltree has `a` or `b` and *not* `c` in the entire ltree". So instead of defining a path, like we do for `lquery`, we're defining a full-text search instead.
    
3. I did not plan to leverage/include the ltree functions in any of my work. I found them mostly useful in stored procedures where you're trying to manipulate ltrees before querying on them and would probably follow if there is any demand down the road.
4. Just like with the functions, I did not plan on including `||` because it seems like out of initial scope for core `ltree` inclusion into Prisma. I don't have much experience with it or it's use cases either.
5. Postgres does recommend leveraging `GiST` indexes on `ltree`s in production however it's not strictly necessary. I would advise we utilize the same Prisma semantics for defining indexes and allow users to specify `GiST` (which I believe is not currently supported).
6. I don't really use any other ORMs, so I can't speak to them but from what I have seen, there aren't many that would probably support `ltree` out of the box.

As far as `ltree` operators are concerned, I have unit tests written in quaint that demonstrate their functionality: https://github.com/halfmatthalfcat/quaint/pull/1/files#diff-dcffd2f91c64a29518e84dac43eb2bfabcf5a146d875d2538b68f8a21dd2455a (you have to expand `src/tests/query.rs` to see them)

**Reply by halfmatthalfcat** [reply]
As far as extensions go, I think it would be great to be able to auto-include them, potentially within the `datasource` block. The issue is, with Postgres at least, a ton of useful extensions must be installed manually and then the extension created via query. `ltree` natively ships with Postgres, so all you need to do is enable it via query which can be done in the initial migration if specified somewhere within Prisma but for others, it's more manual.

There is already precedent though, in Prisma, for relying on the user to make sure they have their db setup correctly; example being with `citext`: https://www.prisma.io/docs/concepts/components/prisma-schema/features-without-psl-equivalent#enable-extensions-for-native-database-functions

**Reply by halfmatthalfcat** [reply]
So I've finally got `ltree`, `lquery` and `ltxtquery` [merged into `rust-postgres`](https://github.com/sfackler/rust-postgres/pull/875), which means we can now update @pimeys's `pgbouncer-mode` fork/branch to bring those changes in quaint.

@pimeys I remember you mentioning earlier that you hope on maintaining a complete fork (or at least, control the postgres lib for Prisma's sake long term) but I'm hoping we can at least update `pgbouncer-mode` in the interim to keep the `ltree` initiative moving.

I have a PR open in your fork: https://github.com/pimeys/rust-postgres/pull/4

**Reply by janpio** [reply]
Nice, seems you already managed with Julius to make progress here and Quaint can now support these three.

Now it gets more interesting: How can we use this in Prisma already? What do you think is the absolute minimum that would make sense for you to surface this? Because it became quite obvious that the _full_ package is a huge thing we really need to understand and then design properly. But maybe something smaller...

**Reply by halfmatthalfcat** [reply]
So yeah, I've been aiming to support `ancestor_of`, `descendant_of`, `matches` and `matches_fulltext` as filters, if `ltree` is a native type on a given field:
- [New filters](https://github.com/halfmatthalfcat/prisma-engines/pull/1/files#diff-273049311f3875c43dd556008ddbc376ce51d2e5224b1c1f1838fb823f15f507R107)
- [Added only if native type is ltree](https://github.com/halfmatthalfcat/prisma-engines/pull/1/files#diff-8493d8e001b02a0db52646d2db5b0513eb606c4c172bb5d7a983448a0fe71b90R159)

I think these 4 concepts, largely, make up the desired functionality of using the `ltree` type. All of the functions and extra things Postgres provides are much more advanced and I would argue, more niche, that can be added at a later time if the demand is there.

Per [the Postgres docs](https://www.postgresql.org/docs/current/ltree.html#LTREE-OP-TABLE) too, `ltree` can support `=, <>, <, >, <=, >=`, so those will be included as well. 

I think this is a good first foray into supporting `ltree`s in Prisma proper. I'm still working on tests and have to go back and update my quaint PR to include the `rust-postgres` bump and add some more documentation.

**Reply by halfmatthalfcat** [reply]
Quaint PR now up: https://github.com/prisma/quaint/pull/352

**Comment by janpio** [comment]
> Appreciate any advice or guidance on how you've managed larger open source contributions to Prisma and what I should expect going forward.

Prisma is very early on our "accept contributions from the community" journey. 

Until now those contributions were mainly in dependent projects lik quaint or tiberius, or on the smaller scale across our TS codebase in `prisma/prisma` or content changes in `prisma/docs`. The reason is what I described above - a) deciding we want to spend time on something as a company and b) then designing a feature that fits the long term strategy is very hard in our current state. 

Prisma did not start as a Open Source project building on community contributions, and as such does not have _any_ of the mechanisms required to make that work. Product Management is internal (Roadmap, Opportunity documents, Solution design, etc), as it Engineering generally (although we work in the open in GitHub, all communication and planning is still behind the scenes in private.

**Reply by halfmatthalfcat** [reply]
Cool, thank you for the context. I completely understand where you're coming from. I can see where it would get hard as, speaking personally, I'm not sure architecturally what the internal standards or plans are for the various codebases. I've been trying to confirm to patterns and conventions currently established (I'm also newer to Rust, so that is another wrinkle) as well as add ample testing to ease the friction/give some reassurance that external code isn't breaking.

I'm not sure I'm going to be very useful in adding suggestions since I've never been in the position of maintaining large OSS codebases but in anyway I can help, I'd be down to lend my opinions/time.

**Reply by janpio** [reply]
You did the best thing someone can do in such a situation already: Reach out to the maintainers early and look for feedback and context. We have on our agenda for this year to understand our situation better and work on ceating a first strategy to (potentially) open up a bit more. Fortunately these things happen around the same time, so we can figure it out together.

**Reply by halfmatthalfcat** [reply]
Cool, so I guess I'll be somewhat of a guinea pig. I'll continue on as-is and we can address process as things progress.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
